File: kata-trivia-python/trivia/game.py
from .player import Player
from .question_deck import QuestionDeck


class Game:
    MAX_PLAYERS = 6
    BOARD_SIZE = 12
    WINNING_COINS = 6
    QUESTIONS_PER_CATEGORY = 50

    def __init__(self):
        self.players = []
        self.questions = QuestionDeck(self.QUESTIONS_PER_CATEGORY)
        self.current_player_index = 0
        self.is_getting_out_of_penalty_box = False

    def is_playable(self):
        return self.how_many_players() >= 2

    def add(self, player_name):
        self._validate_new_player(player_name)
        self.players.append(Player(player_name))

        print(f"{player_name} was added")
        print(f"They are player number {len(self.players)}")
        return True

    def how_many_players(self):
        return len(self.players)

    def roll(self, roll):
        print(f"{self._current_player_name()} is the current player")
        print(f"They have rolled a {roll}")

        if self._current_player().in_penalty_box:
            self._handle_penalty_box_roll(roll)
        else:
            self._take_turn(roll)

    def handle_correct_answer(self):
        if self._current_player().in_penalty_box and not self.is_getting_out_of_penalty_box:
            self._advance_to_next_player()
            return True

        self._award_correct_answer()
        winner = self._did_player_win()
        self._advance_to_next_player()
        return winner

    def wrong_answer(self):
        print("Question was incorrectly answered")
        print(f"{self._current_player_name()} was sent to the penalty box")
        self._current_player().send_to_penalty_box()

        self._advance_to_next_player()
        return True

    def _validate_new_player(self, player_name):
        if len(self.players) == self.MAX_PLAYERS:
            raise ValueError(f"The game cannot have more than {self.MAX_PLAYERS} players")
        if any(player.has_name(player_name) for player in self.players):
            raise ValueError("Player names must be unique")

    def _handle_penalty_box_roll(self, roll):
        if roll % 2 != 0:
            self.is_getting_out_of_penalty_box = True
            self._current_player().release_from_penalty_box()
            print(f"{self._current_player_name()} is getting out of the penalty box")
            self._take_turn(roll)
        else:
            print(f"{self._current_player_name()} is not getting out of the penalty box")
            self.is_getting_out_of_penalty_box = False

    def _take_turn(self, roll):
        self._move_current_player(roll)
        print(f"{self._current_player_name()}'s new location is {self._current_player().position}")
        print(f"The category is {self._current_category()}")
        self._ask_question()

    def _move_current_player(self, roll):
        self._current_player().move(roll, self.BOARD_SIZE)

    def _current_player_name(self):
        return self._current_player().name

    def _current_player(self):
        return self.players[self.current_player_index]

    def _ask_question(self):
        print(self.questions.next_question(self._current_category()))

    def _current_category(self):
        return self.questions.category_for(self._current_player().position)

    def _award_correct_answer(self):
        print("Answer was correct!!!!")
        self._current_player().add_coin()
        print(f"{self._current_player_name()} now has {self._current_player().coins} Gold Coins.")

    def _advance_to_next_player(self):
        self.current_player_index += 1
        if self.current_player_index == len(self.players):
            self.current_player_index = 0

    def _did_player_win(self):
        return not self._current_player().has_won(self.WINNING_COINS)

File: kata-trivia-java/src/main/java/trivia/Game.java
package trivia;

import java.util.ArrayList;
import java.util.List;

// REFACTOR ME
public class Game implements IGame {
   private static final int MAX_PLAYERS = 6;
   private static final int BOARD_SIZE = 12;
   private static final int WINNING_COINS = 6;
   private static final int QUESTIONS_PER_CATEGORY = 50;

   private final List<Player> players = new ArrayList<>();
   private final QuestionDeck questions = new QuestionDeck(QUESTIONS_PER_CATEGORY);

   private int currentPlayer = 0;
   private boolean isGettingOutOfPenaltyBox;

   public boolean isPlayable() {
      return (howManyPlayers() >= 2);
   }

   public boolean add(String playerName) {
      validateNewPlayer(playerName);
      players.add(new Player(playerName));

      System.out.println(playerName + " was added");
      System.out.println("They are player number " + players.size());
      return true;
   }

   private void validateNewPlayer(String playerName) {
      if (players.size() == MAX_PLAYERS) {
         throw new IllegalStateException("The game cannot have more than " + MAX_PLAYERS + " players");
      }
      if (players.stream().anyMatch(player -> player.hasName(playerName))) {
         throw new IllegalArgumentException("Player names must be unique");
      }
   }

   public int howManyPlayers() {
      return players.size();
   }

   public void roll(int roll) {
      System.out.println(currentPlayerName() + " is the current player");
      System.out.println("They have rolled a " + roll);

      if (currentPlayer().inPenaltyBox) {
         handlePenaltyBoxRoll(roll);
      } else {
         takeTurn(roll);
      }
   }

   private void handlePenaltyBoxRoll(int roll) {
      if (roll % 2 != 0) {
         isGettingOutOfPenaltyBox = true;
         currentPlayer().releaseFromPenaltyBox();
         System.out.println(currentPlayerName() + " is getting out of the penalty box");
         takeTurn(roll);
      } else {
         System.out.println(currentPlayerName() + " is not getting out of the penalty box");
         isGettingOutOfPenaltyBox = false;
      }
   }

   private void takeTurn(int roll) {
      moveCurrentPlayer(roll);
      System.out.println(currentPlayerName() + "'s new location is " + currentPlayer().position);
      System.out.println("The category is " + currentCategory());
      askQuestion();
   }

   private void moveCurrentPlayer(int roll) {
      currentPlayer().move(roll, BOARD_SIZE);
   }

   private String currentPlayerName() {
      return currentPlayer().name;
   }

   private Player currentPlayer() {
      return players.get(currentPlayer);
   }

   private void askQuestion() {
      System.out.println(questions.nextQuestion(currentCategory()));
   }


   private String currentCategory() {
      return questions.categoryFor(currentPlayer().position);
   }

   public boolean handleCorrectAnswer() {
      if (currentPlayer().inPenaltyBox && !isGettingOutOfPenaltyBox) {
         advanceToNextPlayer();
         return true;
      }

      awardCorrectAnswer();
      boolean winner = didPlayerWin();
      advanceToNextPlayer();
      return winner;
   }

   private void awardCorrectAnswer() {
      System.out.println("Answer was correct!!!!");
      currentPlayer().addCoin();
      System.out.println(currentPlayerName()
                         + " now has "
                         + currentPlayer().coins
                         + " Gold Coins.");
   }

   public boolean wrongAnswer() {
      System.out.println("Question was incorrectly answered");
      System.out.println(currentPlayerName() + " was sent to the penalty box");
      currentPlayer().sendToPenaltyBox();

      advanceToNextPlayer();
      return true;
   }

   private void advanceToNextPlayer() {
      currentPlayer++;
      if (currentPlayer == players.size()) currentPlayer = 0;
   }


   private boolean didPlayerWin() {
      return !currentPlayer().hasWon(WINNING_COINS);
   }
}
